controllers/dockermachine_controller.go

```diff
@@ -91,137 +132,138 @@ func (r *DockerMachineReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
+	if machine.Spec.Bootstrap.Data == nil {
+		externalMachine.Info("Waiting for the Bootstrap provider controller to set bootstrap data")
 		return ctrl.Result{}, nil
 	}
-}
 
-// State is the state our machine object is in
-type State int
+	// Determine if the machine is the first control plane joining the cluster
+	isFirstControlPlane := util.IsControlPlaneMachine(machine) //&& m.Cluster.Annotations[clusterv1.ClusterAnnotationControlPlaneReady] != "true"
 
-const (
-	// Provisioned is the state that happens after the machine is part of a cluster.
-	// Technically it should be Running but this controller sees no difference between the two.
-	Provisioned State = iota
-	// Pending is when the machine is waiting for bootstrap data to exist.
-	Pending
-	// Deleted is when the machine has been deleted.
-	Deleted
-	// Provisioning is when bootstrap data exists but it's not finished being provisioned.
-	Provisioning
-)
+	//Create the docker container hosting the load balancer
+	if util.IsControlPlaneMachine(machine) {
+		if err := externalMachine.CreateControlPlane(machine.Spec.Version); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to create control plane DockerMachine")
+		}
+	} else {
+		if err := externalMachine.CreateWorker(machine.Spec.Version); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to create worker DockerMachine")
+		}
+	}
 
-// String is a helper function to print a human-readable form of the state.
-func (s State) String() string {
-	switch s {
-	case Provisioning:
-		return "Provisioning"
-	case Deleted:
-		return "Deleted"
-	case Provisioned:
-		return "Provisioned"
-	case Pending:
-		return "Pending"
-	default:
-		return "Unknown"
+	// if the machine is the first control plane added to the cluster, update the load balancer configuration
+	// NB. in case of the first control plane node this step must be executed before kubeadm init
+	if isFirstControlPlane {
+		if err := externalLoadBalancer.UpdateConfiguration(); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to update DockerCluster.loadbalancer configuration")
+		}
 	}
-}
 
-func getState(machine *clusterv1.Machine, dockerMachine *infrav1.DockerMachine) State {
-	// Deleted takes precedence
-	if !machine.ObjectMeta.DeletionTimestamp.IsZero() {
-		return Deleted
+	// exec bootstrap
+	// NB. this step is necessary to mimic the behaviour of cloud-init that is embedded in the base images
+	// for other cloud providers
+	if err := externalMachine.ExecBootstrap(*machine.Spec.Bootstrap.Data); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to exec DockerMachine bootstrap")
 	}
-	if dockerMachine.Spec.ProviderID != nil {
-		return Provisioned
+
+	// if the new machine is an additional control-plane node, update the load balancer configuration
+	// NB. in case of additional control plane nodes this step is executed after kubeadm join in order
+	// to not redirect traffic to an API server instance not yet ready.
```

Comment:
Does the LB not have a health check? As long as there is a health check updating the LB should be fine prior to adding additional control plane instances.

Reply:
Yes the LB has health checks, but while testing parallel node joining in kubeadm we observed some edge conditions still not yet fully clear.
So I would prefer to stay on the safe side...

Reply:
thinking again about this, I prefer to simplify WRT to handling complexity for edge case that can't happen without enabling parallelism for the reconcile loop

Reply:
@fabriziopandini we've already started enabling parallelism to other controllers, adding it to capd will help for testing down the line.